### DISCUSSION TITLE
Advice on using both transports for API flexibility & security

### DISCUSSION THREAD

**Discussion by jd-solanki** [fastapi-users/fastapi-users#1417] [Q&A] [answered]
Hi 👋🏻 

As API can be consumed by anything like Web app or any IoT device, We need cookie transport for web solution and bearer for others so I want to know if developers do provide both transports at the same time?

- If yes, I want to know what convesions do you guys follow? Like endpoint path for both, how do you structure your fastapi-users code?
- If no, Storing non-httpOnly cookie in browser via JS is risky because anyone can steal it so how to you cope with it?

I would like to learn more on flexiblity of Auth API so if someone can give me more details that will be really helpful for me.

Thank you so much 🙏🏻

**Comment by frankie567** [comment, marked as answer]
Yes, you can have several transports (e.g. cookie and bearer) that'll be used depending on the context. The trick is that it'll produce two [authentication backends](https://fastapi-users.github.io/fastapi-users/latest/configuration/authentication/) (a strategy + a transport). So it'll roughly look like this:

```py
cookie_transport = CookieTransport(cookie_max_age=3600) 
bearer_transport = BearerTransport(tokenUrl="/auth/device/login")

def get_jwt_strategy() -> JWTStrategy:
    return JWTStrategy(secret=SECRET, lifetime_seconds=3600)

web_backend = AuthenticationBackend(
    name="web",
    transport=cookie_transport,
    get_strategy=get_jwt_strategy,
)
device_backend = AuthenticationBackend(
    name="device",
    transport=bearer_transport,
    get_strategy=get_jwt_strategy,
)
```

*Side note: you're not even limited to use the same strategy. You could very well have a dedicated strategy for your device backend, like tokens stored in [database](https://fastapi-users.github.io/fastapi-users/latest/configuration/authentication/strategies/database/).*

---

Then, you'll need to wire two different login routers, so you can either create a session using the `web_backend` or the `device_backend`:

```py
fastapi_users = FastAPIUsers[User, uuid.UUID](
    get_user_manager,
    [web_backend, device_backend],  # Notice how we declare the two backends here - It tells FastAPI to check both of them to authenticate an incoming request
)

app = FastAPI()
app.include_router(
    fastapi_users.get_auth_router(web_backend),
    prefix="/auth/web",
    tags=["auth"],
)
app.include_router(
    fastapi_users.get_auth_router(device_backend),
    prefix="/auth/device",
    tags=["auth"],
)
```
